torch/csrc/generic/Storage.cpp

```diff
@@ -31,14 +32,50 @@ static PyObject * THPStorage_(pynew)(PyTypeObject *type, PyObject *args, PyObjec
   THPStoragePtr self((THPStorage *)type->tp_alloc(type, 0));
   THPUtils_assert(self, "failed to allocate a " THPStorageStr " object");
   c10::Allocator* allocator = nullptr;
+//#ifdef USE_CUDA
+//  at::cuda::OptionalCUDAGuard device_guard;
+//#endif
+
+  at::OptionalDeviceGuard device_guard;
 
   // Internally we allow constructing with a keywoard only argument cdata
   if (kwargs != nullptr) {
     PyObject *allocator_ptr = PyDict_GetItemString(kwargs, "allocator");
+    PyObject *device_ptr = PyDict_GetItemString(kwargs, "device");
+
+    TORCH_CHECK(!allocator_ptr || !device_ptr,
+        THPStorageStr, "(): cannot provide both 'allocator' and 'device' kwargs");
+
     if (allocator_ptr) {
       THPUtils_assert(THPUtils_checkLong(allocator_ptr), "invalid allocator");
       allocator = static_cast<c10::Allocator*>(PyLong_AsVoidPtr(allocator_ptr));
       PyDict_DelItemString(kwargs, "allocator");
+
+    } else if (device_ptr) {
+      c10::Device device(at::kCPU);
+      if (THPDevice_Check(device_ptr)) {
```

Comment:
I imagine there's probably be a utility function somewhere that handles turning a Python device, string, or int into a `c10::Device`, but I haven't found it yet. If it doesn't exist, I can add it

Reply:
I think this is what the PythonArgParser does

Reply:
Oh right, I'll just use that then

Reply:
I tried converting this function to use `PythonArgParser`, but there's a small problem. One of the possible positional args to this function is a `PySequence`, but `PythonArgParser` doesn't support that yet. I think there are two options:

* Add support for PySequence
* Or I could just use the type string "PyObject*" and then perform an additional check explicitly within this function to make sure it's a PySequence.

I think the better option is to add support for PySequence. I don't remember exactly how much additional code is required to add a type to PythonArgParser though. I'll find out

Reply:
Meh maybe manual parsing was best lol. Factor out the device parsing in PyArgParser?